Validate recipient email addresses to prevent invalid ending characters

mhf1998 opened the pull request:
Fixes #7112 

- Added a regular expression pattern to identify and invalidate email addresses that end with specific invalid characters.
- The validation logic was incorporated into the `isValidEmail` method, ensuring that any email address ending with characters such as `/`, `?`, `=`, `+`, `&`, `^`, `%`, `$`, `#`, `!`, `'`, and `-` is correctly flagged as invalid. (I checked and all the special characters that led to the repeating of this bug scenario were just these.)
- Improved error handling, and show an error when the focus is changed from the To, CC, and BCC fields; instead of showing the error after sending the email.

![image](https://github.com/user-attachments/assets/8df0dc77-f629-40a5-abef-53e2c671eb53)

cketti reviewed (changes requested):
Please add an additional (in)validity check for the domain part to `UserInputEmailAddressParser` and add relevant tests to `UserInputEmailAddressParserTest`.

Side note: This doesn't fix the underlying problem of us (indirectly) using `Rfc822Tokenizer` which is a bad fit for the job. However, it does improve the current situation a little bit.

mhf1998 commented on legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java:
```diff
@@ -182,13 +184,35 @@ protected Recipient defaultObject(String completionText) {
+    // Method to validate email addresses
+    private boolean isValidEmail(List<Address> addresses) {
+        // Regular expression pattern to identify invalid ending characters in email addresses
+        Pattern invalidEndingPattern = Pattern.compile("[/?=+&^%$#!'\\-]$");
```
> A quick test revealed that adding e.g. a slash in the middle of the domain part is currently also considered valid by `UserInputEmailAddressParser`. Example: `test@domain/example`.
> 
> So only checking the last character of the domain part is not sufficient to catch all instances of this bug. I suggest using `HostNameUtils.isLegalHostNameOrIP()` to validate the domain part of the entered email address.

I read everything you mentioned. Got it. 
I'll get it all done at the first chance I get.

mhf1998 replied:
Done!

mhf1998 commented on legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java:
```diff
@@ -182,13 +184,35 @@ protected Recipient defaultObject(String completionText) {
             }
 
+            // email address validation logic after parsing
+            if (!isValidEmail(parsedAddresses)) {
```
Done!

mhf1998 commented:
I have implemented everything you asked for. Now, the domain is correctly validated by the use of the function you mentioned, and even if the last character of the email is wrong, that is caught as well, and the relevant error is displayed. 
Moreover, all tests have been written and added to the class.

cketti commented on legacy/ui/legacy/src/main/res/values-fa/strings.xml:
```diff
@@ -950,4 +950,5 @@
     <string name="about_app_authors_thunderbird">گروه همراه تاندربرد</string>
-</resources>
\ No newline at end of file
+    <string name="recipient_error_invalid_email">آدرس ایمیل نامعتبر است</string>
```
Translations should only be updated via Weblate. Changing them in a pull request leads to the potential for merge conflicts in Weblate that are annoying to fix.

mhf1998 commented:
@cketti
I have implemented everything you asked for. Done.

cketti commented:
Thanks :+1:

mhf1998 commented:
You're welcome.
It's an honor to collaborate on this project. ☺️
Thank you for taking the time to review and respond.🙏🏻
